File: Assets/scripts/SelectionController.js
#pragma strict
import UnityEngine.UI;
import System.Collections.Generic;

var selectedBots: List.<GameObject> = new List.<GameObject>();

function avatarSelected(avatar: GameObject) {
	var bot = botForAvatar(avatar);
	toggleBotSelection(bot, avatar);
}

function addSelected() {
	insertNewBot();
}

// Avatars
public var avatarPrefab: GameObject;
public var botPrefab: GameObject;

private function insertNewBot() {
	var oldBotCount = bots().length;

	var avatarTransform = Vector3();
	var width = 50.0; // TODO calculate based off space available
	avatarTransform.x = (width + 1.0) * oldBotCount;
	// Debug.Log(avatarTransform);
	var avatar = Instantiate(avatarPrefab, avatarTransform, Quaternion());
	avatar.GetComponent.<Image>().color = colorForSelectionState(false);
	avatar.transform.SetParent(this.transform, false);
	avatar.GetComponent.<Button>().onClick.AddListener(function() {
		avatarSelected(avatar);
	});

	var newBot = Instantiate(botPrefab, Vector3(0, 10, 0), Quaternion());
	newBot.GetComponent.<Renderer>().material.color = colorForSelectionState(false);
}

private function avatars() {
	return GameObject.FindGameObjectsWithTag("avatar");
};

// Bots
function toggleBotSelection(bot: GameObject, avatar: GameObject) {
	var isSelected = selectedBots.Contains(bot);
	if (isSelected) {
		selectedBots.Remove(bot);
	} 
	else {
		selectedBots.Add(bot);
	}
	bot.GetComponent.<Renderer>().material.color = colorForSelectionState(!isSelected);
	avatar.GetComponent.<Image>().color = colorForSelectionState(!isSelected);
}

private function bots() {
	return GameObject.FindGameObjectsWithTag("bot");
};

private function botForAvatar(avatar: GameObject) {
	var index = System.Array.IndexOf(avatars(), avatar);
	return bots()[index];
}

// Conveniences

private function colorForSelectionState(selected): Color {
	return selected ? Color.blue : Color.gray;
}

File: Assets/scripts/SelectionController.cs
using UnityEngine;
using UnityEngine.UI;
using System.Collections.Generic;

public class SelectionController: MonoBehaviour 
{
	void Start() 
	{
		GameObject addButton = this.transform.Find("AddButton").gameObject;
		addButton.GetComponent<Button>().onClick.AddListener(() => insertNewBot());
	}

	public List<GameObject> selectedBots = new List<GameObject>();
	public GameObject avatarPrefab;
	public GameObject botPrefab;

	void avatarSelected(GameObject avatar) 
	{
		GameObject bot = botForAvatar(avatar);
		toggleBotSelection(bot, avatar);
	}

	GameObject botForAvatar(GameObject avatar) 
	{
		int index = System.Array.IndexOf(GameObjectExtensions.AvatarObjects(), avatar);
		return GameObjectExtensions.BotObjects()[index];
	}

	void toggleBotSelection(GameObject bot, GameObject avatar) 
	{
		bool wasSelected = selectedBots.Contains(bot);
		if (wasSelected) {
			selectedBots.Remove(bot);
		} else {
			selectedBots.Add(bot);
		}
		bot.SetSelected(!wasSelected);
		avatar.SetSelected(!wasSelected);
	}

	void insertNewBot() 
	{
		int oldBotCount = GameObjectExtensions.BotObjects().Length;
		Vector3 avatarTransform = new Vector3();
		float width = 50.0f; // TODO calculate based off space available
		avatarTransform.x = (width + 1.0f) * oldBotCount;
		// Debug.Log(avatarTransform);
		GameObject avatar = Instantiate(avatarPrefab, avatarTransform, new Quaternion()) as GameObject;
		avatar.SetSelected(false);
		avatar.transform.SetParent(this.transform, false);
		avatar.GetComponent<Button>().onClick.AddListener(() => avatarSelected(avatar));

		GameObject newBot = Instantiate(botPrefab, new Vector3(0, 10, 0), new Quaternion()) as GameObject;
		newBot.SetSelected(false);
	}
}


